frontendUiTableImpl/src/main/java/com/virtuslab/gitmachete/frontend/ui/impl/table/EnhancedGraphTable.java

```diff
@@ -208,11 +207,76 @@ private void refreshModel(GitRepository gitRepository, List<String> skippedBranc
   }
 
   @UIEffect
-  private void openDiscoverDialog() {
-    var action = ActionManager.getInstance().getAction(ACTION_DISCOVER);
-    var dataContext = DataManager.getInstance().getDataContext(EnhancedGraphTable.this);
-    var anActionEvent = AnActionEvent.createFromDataContext(ACTION_PLACE_EMPTY_TABLE, new Presentation(), dataContext);
-    action.actionPerformed(anActionEvent);
+  private void doAutomaticDiscover(Path macheteFilePath) {
+    var selectedRepository = getGitRepositorySelectionProvider().getSelectedGitRepository().getOrNull();
+    if (selectedRepository == null) {
+      LOG.error("Can't do automatic discover because of undefined selected repository");
+      return;
+    }
+    var mainDirPath = GitVfsUtils.getMainDirectoryPath(selectedRepository).toAbsolutePath();
+    var gitDirPath = GitVfsUtils.getGitDirectoryPath(selectedRepository).toAbsolutePath();
+
+    // `nullness` to allow runResult be uninitialized and IBranchLayoutWriter#write accept @Nullable IBranchLayout without redundant check
+    // `guieffect` to allow running lambdas that are @AlwaysSafe in methods onSuccess and onFailure
+    @SuppressWarnings({"nullness", "guieffect"}) var task = new Task.Backgroundable(project,
+        getString("string.GitMachete.EnhancedGraphTable.automatic-discover.task-title")) {
+      private Try<DiscoverResult> discoverResult;
+
+      @Override
+      public void run(ProgressIndicator indicator) {
+        discoverResult = Try.of(() -> RuntimeBinding.instantiateSoleImplementingClass(IGitMacheteRepositoryCache.class)
+            .getInstance(mainDirPath, gitDirPath).discoverLayoutAndCreateSnapshot())
+            .filterTry(repositorySnapshot -> repositorySnapshot.getRootBranches().size() > 0,
+                __ -> new EmptyRepositoryException())
+            .map(repositorySnapshot -> {
+              var branchLayoutWriter = project.getService(BranchLayoutWriterProvider.class).getBranchLayoutWriter();
+              var branchLayout = repositorySnapshot.getBranchLayout().getOrNull();
+              return DiscoverResult.of(branchLayoutWriter, branchLayout, repositorySnapshot);
+            })
+            .filterTry(result -> result.branchLayout != null, __ -> new EmptyBranchLayoutException())
+            .andThenTry(
+                result -> result.branchLayoutWriter.write(macheteFilePath, result.branchLayout, /* backupBranchLayout */ true));
+      }
+      // This method is invoked on UI dispatch thread
+      @UIEffect
+      @Override
+      public void onSuccess() {
+        discoverResult.onFailure(exception -> {
+          Class<? extends Throwable> exceptionClass = exception.getClass();
```

Comment:
Hmm no idea why this failing on `FenumChecker` :thinking: anyway, possibly as in the[comment](https://github.com/VirtusLab/git-machete-intellij-plugin/pull/550/files#r478652551) above, the temporary exception classes can be eliminated

Reply:
get rid of this